transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -229,6 +229,11 @@ public void setIoRatio(int ioRatio) {
         this.ioRatio = ioRatio;
     }
 
+    @Override
+    public int registeredChannels() {
+        return channels.size();
```

Comment:
I think this does not really work as `channels` is not thread-safe.

Reply:
I'd imagine maintaining a separate (volatile) counter for the number of channels involves some overhead that we could possible avoid accepting that the numbers may not be super precise. This is, perhaps, not a big deal for observability but can surely hurt if we decide to load-balance channels based on it.

What do you think about hoisting such counter under `SingleThreadEvelLoop`; increment it on `register` and decrement it on `channel.closePromise` resolution?

Reply:
@vkostyukov the problem here is that it is possible that a `Channel` will be deregistered but not closed. Also a `Channel` may be re-registered to an other EventLoop.

Reply:
@vkostyukov did you see this ?

Reply:
@normanmaurer @vkostyukov `channels` are stored in `IntObjectHashMap`, as for method `size()`, i think it will be ok even if the `IntObjectHashMap` is not thread safe, because sometimes we don't need the `size()` to be very precise. Maybe we can change the `registeredChannels ` method name or add some information in the doc ?

Reply:
Right. This operation is pretty racy by nature and maybe it's fine if we're not super precise (as long as we document it). I'd imagine these pointers will be flushed pretty quickly on modern hardware anyway.

Happy to add some JavaDoc here explaining the trade off.

Reply:
@vkostyukov that's "ok" with me... That said I think we should "fix" this for netty 5 fundamentally. See https://github.com/netty/netty/pull/8680